File: core/app-server/model/scheduled_task.go
package model

import (
	"encoding/json"
	"time"

	"gorm.io/gorm"
)

// ScheduledTask 定时任务
// schedule_type: atime=指定时间执行一次, cron=cron表达式重复, every=每N秒执行一次
type ScheduledTask struct {
	ID        int64          `json:"id" gorm:"primaryKey;autoIncrement"`
	CreatedAt time.Time      `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt time.Time      `json:"updated_at" gorm:"autoUpdateTime"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	User            string          `json:"user" gorm:"size:255;not null;index;comment:归属用户"`
	App             string          `json:"app" gorm:"size:255;not null;index;comment:归属应用"`
	Name            string          `json:"name" gorm:"size:255;comment:任务名称（用户可编辑）"`
	FullCodePath    string          `json:"full_code_path" gorm:"size:500;not null;comment:要执行的表单路径"`
	Action          string          `json:"action" gorm:"size:32;default:execute;comment:执行动作：execute/table_create/table_update/table_delete"`
	Method          string          `json:"method" gorm:"size:10;default:POST;comment:GET/POST"`
	Payload         json.RawMessage `json:"payload" gorm:"type:json;comment:请求体"`
	RequestUser     string          `json:"request_user" gorm:"size:255;comment:以谁的身份执行"`
	RequestUserDept string          `json:"request_user_dept" gorm:"size:500;comment:请求用户部门路径（透传用）"`
	CreatedBy       string          `json:"created_by" gorm:"size:255;index;comment:创建人（用于列表筛选）"`

	ScheduleType    string     `json:"schedule_type" gorm:"size:20;not null;index;comment:atime/cron/every"`
	RunAt           time.Time  `json:"run_at" gorm:"comment:atime 的执行时间；cron/every 的创建生效时间"`
	NextRunAt       *time.Time `json:"next_run_at" gorm:"index;comment:下次执行时间"`
	LeaseOwner      string     `json:"lease_owner" gorm:"size:128;index;comment:执行租约持有者"`
	LeaseUntil      *time.Time `json:"lease_until" gorm:"index;comment:执行租约到期时间"`
	CronExpr        string     `json:"cron_expr" gorm:"size:100;comment:cron表达式"`
	IntervalSeconds int64      `json:"interval_seconds" gorm:"comment:every 时间间隔秒"`
	MaxRuns         int        `json:"max_runs" gorm:"default:0;comment:every 最多执行次数,0不限制"`
	RunCount        int        `json:"run_count" gorm:"default:0;comment:已执行次数"`

	Status       string `json:"status" gorm:"size:20;not null;index;comment:pending/done/failed/cancelled"`
	Timezone     string `json:"timezone" gorm:"size:64;comment:时区"`
	ErrorMessage string `json:"error_message" gorm:"type:text;comment:最近一次失败信息"`
}

func (ScheduledTask) TableName() string {
	return "scheduled_task"
}

File: web/src/architecture/infrastructure/api/scheduledTask.ts
/**
 * 定时任务 API
 */

import { get, post, del } from '@/architecture/infrastructure/apiClient/request'

/** 与后端一致：execute=普通函数；table_*=表格增改删（走 /_callback） */
export type ScheduledTaskAction =
  | 'execute'
  | 'table_create'
  | 'table_update'
  | 'table_delete'

export type ScheduledTaskNotifyOn = 'none' | 'all' | 'success' | 'failed'

export interface CreateScheduledTaskReq {
  name: string
  full_code_path: string
  action?: ScheduledTaskAction
  method?: string
  payload?: Record<string, unknown> | string
  request_user?: string
  request_user_dept?: string
  schedule_type: 'atime' | 'cron' | 'every'
  run_at?: string
  cron_expr?: string
  interval_seconds?: number
  max_runs?: number
  timezone?: string
  notify_users?: string[]
  notify_departments?: string[]
  notify_on?: ScheduledTaskNotifyOn
}

export interface ScheduledTaskItem {
  id: number
  name: string
  user: string
  app: string
  full_code_path: string
  action?: string
  method?: string
  payload?: string
  request_user?: string
  request_user_dept?: string
  created_by: string
  schedule_type: string
  run_at: string
  next_run_at?: string
  cron_expr?: string
  interval_seconds?: number
  max_runs?: number
  timezone?: string
  status: string
  run_count: number
  error_message?: string
  notify_users?: string[]
  notify_departments?: string[]
  notify_on?: ScheduledTaskNotifyOn
  created_at: string
}

export interface ScheduledTaskExecutionItem {
  id: number
  task_id: number
  executed_at: string
  status: string
  duration_millis?: number
  request_payload: string
  response_payload: string
  error_message?: string
  trace_id?: string
  created_at: string
}

export interface ListScheduledTasksResp {
  list: ScheduledTaskItem[]
  total: number
}

export interface ListScheduledTaskExecutionsResp {
  list: ScheduledTaskExecutionItem[]
  total: number
}

export function methodForScheduledTaskAction(action: ScheduledTaskAction = 'execute', fallback = 'POST'): string {
  switch (action) {
    case 'table_create':
      return 'POST'
    case 'table_update':
      return 'PUT'
    case 'table_delete':
      return 'DELETE'
    default:
      return fallback || 'POST'
  }
}

export function createScheduledTask(data: CreateScheduledTaskReq): Promise<ScheduledTaskItem> {
  // 直接传对象，后端存成 json.RawMessage 即 {"a":1} 的字节；不要 JSON.stringify(payload) 否则会变成字符串导致执行时 unmarshal 报错
  const payload = data.payload ?? {}
  const action = data.action ?? 'execute'
  return post<ScheduledTaskItem>('/workspace/api/v1/scheduled_tasks', {
    name: data.name,
    full_code_path: data.full_code_path,
    action,
    method: data.method || methodForScheduledTaskAction(action),
    payload,
    request_user: data.request_user,
    request_user_dept: data.request_user_dept,
    schedule_type: data.schedule_type,
    run_at: data.run_at,
    cron_expr: data.cron_expr,
    interval_seconds: data.interval_seconds,
    max_runs: data.max_runs ?? 0,
    timezone: data.timezone,
    notify_users: data.notify_users ?? [],
    notify_departments: data.notify_departments ?? [],
    notify_on: data.notify_on ?? 'none'
  })
}

export function getScheduledTask(id: number): Promise<ScheduledTaskItem> {
  return get<ScheduledTaskItem>(`/workspace/api/v1/scheduled_tasks/${id}`)
}

/** full_code_path 为前缀：返回该路径及子路径下的任务（目录节点可看到子表单的定时任务） */
export function listScheduledTasks(params: {
  full_code_path?: string
  status?: string
  page?: number
  page_size?: number
}): Promise<ListScheduledTasksResp> {
  return get<ListScheduledTasksResp>('/workspace/api/v1/scheduled_tasks', params)
}

export function cancelScheduledTask(id: number): Promise<void> {
  return post<void>(`/workspace/api/v1/scheduled_tasks/${id}/cancel`)
}

export function deleteScheduledTask(id: number): Promise<void> {
  return del<void>(`/workspace/api/v1/scheduled_tasks/${id}`)
}

export function listScheduledTaskExecutions(
  taskId: number,
  params?: { status?: string; page?: number; page_size?: number }
): Promise<ListScheduledTaskExecutionsResp> {
  return get<ListScheduledTaskExecutionsResp>(
    `/workspace/api/v1/scheduled_tasks/${taskId}/executions`,
    params
  )
}

export function getScheduledTaskExecution(taskId: number, executionId: number): Promise<ScheduledTaskExecutionItem> {
  return get<ScheduledTaskExecutionItem>(
    `/workspace/api/v1/scheduled_tasks/${taskId}/executions/${executionId}`
  )
}
